From: Danny Feng <dfeng@redhat.com>
Date: Thu, 24 Feb 2011 10:29:36 -0500
Subject: [fs] partitions: Validate map_count in Mac part tables
Message-id: <20110224102936.29136.54785.sendpatchset@dhcp-65-98.nay.redhat.com>
Patchwork-id: 33602
O-Subject: [PATCH RHEL5.7] fs/partitions: Validate map_count in Mac partition
	tables
Bugzilla: 679284
CVE: CVE-2011-1010

RHBZ#:
https://bugzilla.redhat.com/show_bug.cgi?id=679284

backport of:

commit fa7ea87a057958a8b7926c1a60a3ca6d696328ed
Author: Timo Warns <warns@pre-sense.de>
Date:   Thu Feb 17 22:27:40 2011 +0100

    fs/partitions: Validate map_count in Mac partition tables

    Validate number of blocks in map and remove redundant variable.

    Signed-off-by: Timo Warns <warns@pre-sense.de>
    Cc: stable@kernel.org
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Resolves CVE-2011-1010

diff --git a/fs/partitions/mac.c b/fs/partitions/mac.c
index c087100..2f18297 100644
--- a/fs/partitions/mac.c
+++ b/fs/partitions/mac.c
@@ -29,10 +29,9 @@ static inline void mac_fix_string(char *stg, int len)
 
 int mac_partition(struct parsed_partitions *state, struct block_device *bdev)
 {
-	int slot = 1;
 	Sector sect;
 	unsigned char *data;
-	int blk, blocks_in_map;
+	int slot, blocks_in_map;
 	unsigned secsize;
 #ifdef CONFIG_PPC_PMAC
 	int found_root = 0;
@@ -59,10 +58,14 @@ int mac_partition(struct parsed_partitions *state, struct block_device *bdev)
 		put_dev_sector(sect);
 		return 0;		/* not a MacOS disk */
 	}
-	printk(" [mac]");
 	blocks_in_map = be32_to_cpu(part->map_count);
-	for (blk = 1; blk <= blocks_in_map; ++blk) {
-		int pos = blk * secsize;
+	if (blocks_in_map < 0 || blocks_in_map >= MAX_PART) {
+		put_dev_sector(sect);
+		return 0;
+	}
+	printk(" [mac]");
+	for (slot = 1; slot <= blocks_in_map; ++slot) {
+		int pos = slot * secsize;
 		put_dev_sector(sect);
 		data = read_dev_sector(bdev, pos/512, &sect);
 		if (!data)
@@ -111,13 +114,11 @@ int mac_partition(struct parsed_partitions *state, struct block_device *bdev)
 			}
 
 			if (goodness > found_root_goodness) {
-				found_root = blk;
+				found_root = slot;
 				found_root_goodness = goodness;
 			}
 		}
 #endif /* CONFIG_PPC_PMAC */
-
-		++slot;
 	}
 #ifdef CONFIG_PPC_PMAC
 	if (found_root_goodness)
